chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -13,60 +13,94 @@
-        timelock::OutputTimeLock, AccountCommand, AccountSpending, ChainConfig, PoolId, TxInput,
-        TxOutput,
+        output_value::OutputValue,
+        timelock::OutputTimeLock,
+        tokens::{TokenData, TokenId, TokenIssuanceVersion},
+        AccountCommand, AccountSpending, ChainConfig, DelegationId, PoolId, Transaction, TxInput,
+        TxOutput, UtxoOutPoint,
     },
-    primitives::{Amount, BlockDistance, BlockHeight},
+    primitives::{Amount, BlockHeight, Id},
 };
 use utils::ensure;
 
+use crate::{transaction_verifier::amounts_map::CoinOrTokenId, Fee};
+
 use super::IOPolicyError;
 
+type BlockCount = u64;
+
 /// `ConstrainedValueAccumulator` helps avoiding messy inputs/outputs combinations analysis by
 /// providing a set of properties that should be satisfied. For example instead of checking that
 /// all outputs are timelocked when the pool is decommissioned `ConstrainedValueAccumulator` gives a way
 /// to check that an accumulated output value is locked for sufficient amount of time which allows
 /// using other valid inputs and outputs in the same tx.
-///
-/// TODO: this struct can be extended to collect tokens replacing `AmountsMap`
 pub struct ConstrainedValueAccumulator {
-    unconstrained_value: Amount,
-    timelock_constrained: BTreeMap<BlockDistance, Amount>,
+    unconstrained_value: BTreeMap<CoinOrTokenId, Amount>,
+    timelock_constrained: BTreeMap<BlockCount, Amount>,
 }
 
 impl ConstrainedValueAccumulator {
     pub fn new() -> Self {
         Self {
-            unconstrained_value: Amount::ZERO,
+            unconstrained_value: Default::default(),
             timelock_constrained: Default::default(),
         }
     }
 
-    /// Return accumulated amounts that are left
-    // TODO: for now only used in tests but can be used to calculate fees
-    #[allow(dead_code)]
-    pub fn consume(self) -> Result<Amount, IOPolicyError> {
-        self.timelock_constrained
-            .into_values()
+    /// Return accumulated coins that are left
+    pub fn consume(
+        self,
+        chain_config: &ChainConfig,
+        block_height: BlockHeight,
+    ) -> Result<Fee, IOPolicyError> {
+        let unconstrained_change = self
+            .unconstrained_value
+            .get(&CoinOrTokenId::Coin)
+            .cloned()
+            .unwrap_or(Amount::ZERO);
+
+        let maturity_distance: BlockCount = chain_config
+            .staking_pool_spend_maturity_distance(block_height)
+            .to_int()
+            .try_into()
+            .map_err(|_| IOPolicyError::NegativeMaturityDistance)?;
+
+        let timelocked_change = self
+            .timelock_constrained
+            .into_iter()
+            .filter_map(|(lock, amount)| (lock <= maturity_distance).then_some(amount))
             .sum::<Option<Amount>>()
-            .and_then(|v| v + self.unconstrained_value)
-            .ok_or(IOPolicyError::AmountOverflow)
+            .ok_or(IOPolicyError::CoinOrTokenOverflow(CoinOrTokenId::Coin))?;
```

Comment:
I'm wondering if including timelocked coins does not bypass the point of the timelock. Some coins the owner would have to normally wait to be able to spend could be spent immediately if used to pay fees. That does not sound right, unless there are some considerations I missed.

Reply:
The fee is not available to the staker because it is distributed and added to the pool balance which cannot be spent without timelock constraints.